docker/test/init-cluster.sh

```diff
@@ -0,0 +1,250 @@
+			  --host=$(echo $MONGODB_ADDRESS | cut -d: -f1) \
+			  --port=$(echo $MONGODB_ADDRESS | cut -d: -f2) \
+				--eval='db.createUser({
+					user: "'${TEST_MONGODB_USERNAME}'",
+					pwd: "'${TEST_MONGODB_PASSWORD}'",
+					roles: [
+						{ db: "admin", role: "backup" },
+						{ db: "admin", role: "clusterMonitor" },
+						{ db: "admin", role: "restore" },
+						{ db: "config", role: "read" },
+						{ db: "test", role: "readWrite" }
+					]
+				})' \
+				admin
+			[ $? == 0 ] && echo "# INFO: added test user to ${MONGODB_ADDRESS}" && break
+		fi
+		echo "# INFO: retrying db.createUser() on ${MONGODB_ADDRESS} in $sleep_secs secs (try $tries/$max_tries)"
+		sleep $sleep_secs
+		tries=$(($tries + 1))
+	done
+done
+echo "# INFO: all replsets have auth user(s)"
+
+
+shard1=${TEST_MONGODB_S1_RS}'/s1-mongo1:27018,s1-mongo2:27018'
+shard2=${TEST_MONGODB_S2_RS}'/s2-mongo1:27018,s2-mongo2:27018'
+for shard in $shard1 $shard2; do
+	tries=1
+	while [ $tries -lt $max_tries ]; do
+		ADDSHARD=$(/usr/bin/mongo ${MONGO_FLAGS} \
+			--username=${TEST_MONGODB_ADMIN_USERNAME} \
+			--password=${TEST_MONGODB_ADMIN_PASSWORD} \
+			--host=mongos \
+			--port=27017 \
+			--eval='printjson(sh.addShard("'$shard'").ok)' \
+			admin 2>/dev/null)
+		[ $? == 0 ] && [ "$ADDSHARD" == "1" ] && break
+		echo "# INFO: retrying sh.addShard() check for '$shard' in $sleep_secs secs (try $tries/$max_tries)"
+		sleep $sleep_secs
+		tries=$(($tries + 1))
+	done
+	if [ $tries -ge $max_tries ]; then
+		echo "# ERROR: reached max tries $max_tries for '$shard', exiting"
+		exit 1
+	fi
+	echo "# INFO: added shard: $shard"
+done
+
+tries=1
+while [ $tries -lt $max_tries ]; do
+	ENABLESHARDING=$(/usr/bin/mongo ${MONGO_FLAGS} \
+		--username=${TEST_MONGODB_ADMIN_USERNAME} \
+		--password=${TEST_MONGODB_ADMIN_PASSWORD} \
+		--host=mongos \
+		--port=27017 \
+		--eval='sh.enableSharding("test").ok' \
+		admin 2>/dev/null)
+	[ $? == 0 ] && [ "$ENABLESHARDING" == "1" ] && break
+	echo "# INFO: retrying sh.enableSharding(\"test\") check in $sleep_secs secs (try $tries/$max_tries)"
+	sleep $sleep_secs
+	tries=$(($tries + 1))
+done
+if [ $tries -ge $max_tries ]; then
+	echo "# ERROR: reached max tries $max_tries for '$shard', exiting"
+	exit 1
+fi
+echo "# INFO: \"test\" database is now sharded"
+
+tries=1
+while [ $tries -lt $max_tries ]; do
+	SHARDCOL=$(/usr/bin/mongo ${MONGO_FLAGS} \
+		--username=${TEST_MONGODB_ADMIN_USERNAME} \
+		--password=${TEST_MONGODB_ADMIN_PASSWORD} \
+		--host=mongos \
+		--port=27017 \
+		--eval='sh.shardCollection("test.test", {_id: 1}).ok' \
+		admin 2>/dev/null)
+	[ $? == 0 ] && [ "$ENABLESHARDING" == "1" ] && break
+	echo "# INFO: retrying sh.shardCollection(\"test.test\", {_id: 1}) check in $sleep_secs secs (try $tries/$max_tries)"
+	sleep $sleep_secs
+	tries=$(($tries + 1))
+done
+if [ $tries -ge $max_tries ]; then
+	echo "# ERROR: reached max tries $max_tries for '$shard', exiting"
+	exit 1
+fi
+echo "# INFO: \"test.test\" collection is now sharded"
```

Comment:
IMHO, a Go version of that code will be easier to support.

Reply:
Probably yes, but I don't sure it will be easy to write it. I've ported it from `percona-toolkit` just to make it works. 

In the next iteration, I'm going to make at least a lot of comments in this file that explain what is going on there. Then I may try to rewrite it to go but I don't really sure it worth.

In general, I don't really like using compiled languages in build scripts because they can have a lot of not needed dependencies and also because it needs to compile them.

My POV is the best solution for building/initialization/automation stuff is using `bash+make`. Sometimes it hard to understand but it works and everybody can check it anywhere.